block_cascade/cli/crud.py

```diff
@@ -0,0 +1,259 @@
+import click
+import json
+from google.cloud import aiplatform_v1beta1 as aiplatform
+from google.protobuf import json_format
+
+from block_cascade import GcpResource
+from block_cascade.config import find_default_configuration
+from block_cascade.consts import SERVICE
+from block_cascade.utils import get_gcloud_config
+
+
+def get_gcp_project(config: dict) -> str:
+    """Return the GCP project from a local config."""
+    return config.get("core", {})["project"]
+
+
+def get_gcp_region(config: dict) -> str:
+    """Return the GCP region from a local config."""
+    return config.get("compute", {})["region"]
+
+
+def get_persistent_resource_payload(resource: GcpResource) -> dict:
+    """Generate a persistent resource payload from a GcpResource."""
+
+    CPU_MACHINE_TYPE = resource.chief.type
+    CPU_REPLICA_COUNT = resource.chief.count
+    CPU_MIN_REPLICA_COUNT = resource.chief.min_replica_count or 0
+    CPU_MAX_REPLICA_COUNT = resource.chief.max_replica_count or 0
+    if resource.chief.accelerator is not None:
+        GPU_MACHINE_TYPE = resource.chief.type
+        GPU_REPLICA_COUNT = resource.chief.count
+        GPU_ACCELERATOR_TYPE = resource.chief.accelerator.type
+        GPU_ACCELERATOR_COUNT = resource.chief.accelerator.count
+    else:
+        GPU_MACHINE_TYPE = None
+        GPU_REPLICA_COUNT = 0
+        GPU_ACCELERATOR_TYPE = None
+        GPU_ACCELERATOR_COUNT = 0
+
+    # This is consistent with the default disk spec of jobs.
+    DISK_SPEC = {"boot_disk_type": "pd-ssd", "boot_disk_size_gb": 100}
+
+    CPU_AUTOSCALING_SPEC = {
+        "min_replica_count": CPU_MIN_REPLICA_COUNT,
+        "max_replica_count": CPU_MAX_REPLICA_COUNT,
+    }
+
+    CPU_POOL = {
+        "machine_spec": {
+            "machine_type": CPU_MACHINE_TYPE,
+        },
+        "replica_count": CPU_REPLICA_COUNT,
+        "disk_spec": DISK_SPEC,
+    }
+    if CPU_MIN_REPLICA_COUNT > 0 or CPU_MAX_REPLICA_COUNT > 0:
+        CPU_POOL["autoscaling_spec"] = CPU_AUTOSCALING_SPEC
+
+    GPU_POOL = {
+        "machine_spec": {
+            "machine_type": GPU_MACHINE_TYPE,
+            "accelerator_type": GPU_ACCELERATOR_TYPE,
+            "accelerator_count": GPU_ACCELERATOR_COUNT,
+        },
+        "replica_count": GPU_REPLICA_COUNT,
+        "disk_spec": DISK_SPEC,
+    }
+
+    if CPU_REPLICA_COUNT > 0 and GPU_REPLICA_COUNT > 0:
+        RESOURCE_POOLS = [CPU_POOL, GPU_POOL]
+    elif CPU_REPLICA_COUNT > 0:
+        RESOURCE_POOLS = [CPU_POOL]
+    elif GPU_REPLICA_COUNT > 0:
+        RESOURCE_POOLS = [GPU_POOL]
+    else:
+        RESOURCE_POOLS = []
+
+    PERSISTENT_RESOURCE = {
+        "display_name": resource.persistent_resource_id,
+        "resource_pools": RESOURCE_POOLS,
+    }
```

Comment:
Probably should err out if `if not RESOURCE_POOLS` (why would one create empty resource pools).